### PR TITLE
[dattri.algorithm] Fix bugs in TRAK projection

### PR DESCRIPTION
**Pull request by sx-liu**
### 1. Motivation and Context

Current TRAK projection fails for large models. Since all the parameters are flattened and concatenated together, as long as the total number of parameters is larger than `max_chunk_size`, the projector will fail to arrange inputs correctly and raise an internal error.

### 2. Summary of the change

This pr changes the inputs passed to the projector from a large tensor to a dictionary and add necessary condition check for the input model.

Specifically, modified `TRAKAttributor` in `dattri.algorithm.trak`

### 3. What tests have been added/updated for the change?
- [x] N/A: No test will be added
This error only occurs for large models (at least > 0.1b). The error is not caused by any problems in the modules, but by the input format.

### CODE REVIEW

**Comment by tingwl0122** [comment]
Thanks for the PR, @sx-liu!
Could you also verify the performance of this TRAK version?
I assume you are attributing some larger models, does the performance (i.e., LDS) look normal for you task?

**Comment by sx-liu** [comment]
> Thanks for the PR, @sx-liu! Could you also verify the performance of this TRAK version? I assume you are attributing some larger models, does the performance (i.e., LDS) look normal for you task?

I haven't tried it yet, but after running the example on mnist with some small models (lr), I found the accuracy is pretty close to the previous results.

**Comment by tingwl0122** [comment]
> > Thanks for the PR, @sx-liu! Could you also verify the performance of this TRAK version? I assume you are attributing some larger models, does the performance (i.e., LDS) look normal for you task?
> 
> I haven't tried it yet, but after running the example on mnist with some small models (lr), I found the accuracy is pretty close to the previous results.
> 
> ```
> Peak memory usage: 36.384256 MB
> torch.Size([1000])
> [(0, 0), (100, 78), (200, 100), (300, 101), (400, 102), (500, 102), (600, 102), (700, 103), (800, 103), (900, 103)]
> Checked Data Sample      Found flipped Sample     
> --------------------------------------------------
> 0                        0                        
> 100                      78                       
> 200                      100                      
> 300                      101                      
> 400                      102                      
> 500                      102                      
> 600                      102                      
> 700                      103                      
> 800                      103                      
> 900                      103                      
> ```
> 
> I will try to verify it with MT5 model on a small scale of experiment.

yeah, but if this model is not large enough, then the desired functionality will not be tested right

File: dattri/algorithm/trak.py [hunk 1]
```diff
@@ -108,6 +125,21 @@ def cache(
...
+                    for g in grad_t
+                ]
+
+                # Stack the gradient tensors for each layer
+                grad_t = {
+                    key: torch.stack(
+                        [g[key] for g in grad_t],
+                        ).view(len(grad_t), -1)
+                    for key in grad_t[0]
+                }
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:141]: I don't think we need this step? Eventually all the params (in each module) will be projected and added together.

File: dattri/algorithm/trak.py [hunk 2]
```diff
@@ -210,6 +242,20 @@ def attribute(
...
+                        for g in grad_t
+                    ]
+
+                    # Stack the gradient tensors for each layer
+                    grad_t = {
+                        key: torch.stack(
+                            [g[key] for g in grad_t],
+                            ).view(len(grad_t), -1)
+                        for key in grad_t[0]
+                    }
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:257]: same issue as above

File: dattri/algorithm/trak.py [hunk 3]
```diff
@@ -245,6 +291,20 @@ def attribute(
...
+                    for g in grad_t
+                ]
+
+                # Stack the gradient tensors for each layer
+                grad_t = {
+                    key: torch.stack(
+                        [g[key] for g in grad_t],
+                        ).view(len(grad_t), -1)
+                    for key in grad_t[0]
+                }
```
Comment (tingwl0122) [inline dattri/algorithm/trak.py:306]: same issue as above

**Review by tingwl0122 — COMMENTED** [review]
If it is hard to get real LDS performance, we can just make sure `Chunkedcudaprojector` is called when we are attributing big models.

File: dattri/algorithm/trak.py [hunk 4]
```diff
@@ -71,6 +75,19 @@ def m(params, image_label_pair):
...
+                    self.projector_kwargs["proj_max_batch_size"])
+        for p in self.task.model.parameters():
+            num_params = p.numel()
+            if num_params > max_chunk_size:
+                error_msg = (f"Your model contains layers with at least {num_params} "
+                             "parameters, which is larger than the `max_chunk_size` "
+                             f"{max_chunk_size} of the projector. This is currently "
+                             "not supported. To avoid this error, please try a smaller"
+                             " `proj_max_batch_size` or use cpu projector instead.")
+                raise ValueError(error_msg)
```
[the remark this answers is not in the document]
Comment (sx-liu) [reply dattri/algorithm/trak.py:89]: Actually even if all the inputs are passed as dict there is still an upper bound on the number of parameters in each single layer. Some layers, say the embedding layers in LLMs, can be even larger the chunk size of some projectors.

 We could remove this error after we implemented the function of further breaking the tensors down into feasible sizes.

File: dattri/algorithm/trak.py (hunk 1 above)
Comment (sx-liu) [reply dattri/algorithm/trak.py:141]: I think it's just a convenient way to project batched gradients. Here we just stacked the gradients in each layer as `(batch_size, dim_out * dim_in)` so that the projector can process them all at once. Otherwise we may need to use a loop to process every gradient in the batch one by one.

**Comment by sx-liu** [comment]
> > > Thanks for the PR, @sx-liu! Could you also verify the performance of this TRAK version? I assume you are attributing some larger models, does the performance (i.e., LDS) look normal for you task?
> > 
> > 
> > I haven't tried it yet, but after running the example on mnist with some small models (lr), I found the accuracy is pretty close to the previous results.
> > ```
> > Peak memory usage: 36.384256 MB
> > torch.Size([1000])
> > [(0, 0), (100, 78), (200, 100), (300, 101), (400, 102), (500, 102), (600, 102), (700, 103), (800, 103), (900, 103)]
> > Checked Data Sample      Found flipped Sample     
> > --------------------------------------------------
> > 0                        0                        
> > 100                      78                       
> > 200                      100                      
> > 300                      101                      
> > 400                      102                      
> > 500                      102                      
> > 600                      102                      
> > 700                      103                      
> > 800                      103                      
> > 900                      103                      
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I will try to verify it with MT5 model on a small scale of experiment.
> 
> yeah, but if this model is not large enough, then the desired functionality will not be tested right

I have tested MT5 model on ftrace dataset and obtained the MRR. The MRR is approximately 0.25 with only one checkpoint. The value is lower than the released one in the paper but I think it's reasonable because their experiment used maybe 10 checkpoints.

And I think the experiment on MNIST can also demonstrate that at least the unflattening of gradient is correct, because whichever projector we use the unflattening added in this PR is the same.

**Comment by tingwl0122** [comment]
Will not follow this direction since unflattening will still cause problems if a single layer is too large (i.e. exceed the maximum chunk size).
We will have another PR to always concat the grads to tensors and split the tensors into pieces for projection if needed.
